File: pegasusX/apps/retailer-app-ios/retailerapp/retailerapp/Design/StatusStack.swift
import SwiftUI

enum StatusStackMode: String {
    case empty
    case zero
    case live
    case unavailable
}

struct StatusStackRow: Equatable {
    let key: String
    let count: Int?
    let share: Double
}

struct StatusStackModel: Equatable {
    let mode: StatusStackMode
    let rows: [StatusStackRow]
    let total: Int
}

let orderStatusFunnel = [
    "PENDING", "SCHEDULED", "AUTO_ACCEPTED", "BACKORDERED",
    "LOADED", "IN_TRANSIT", "DELAYED",
    "ARRIVED", "ARRIVED_SHOP_CLOSED",
    "AWAITING_PAYMENT", "PENDING_CASH_COLLECTION", "DELIVERED_ON_CREDIT",
    "FISCALIZING", "FISCAL_FAILED", "RECONCILIATION_REQUIRED",
    "COMPLETED", "CANCELLED",
]

let manifestStates = [
    "DRAFT", "LOADING", "SEALED", "DISPATCHED", "COMPLETED", "CANCELLED",
]

let truckDutyStatuses = [
    "AVAILABLE",
    "IN_TRANSIT",
    "RETURNING_TO_WAREHOUSE",
    "OFF_SHIFT",
    "UNASSIGNED",
    "VEHICLE_INACTIVE",
    "UNAVAILABLE",
    "INACTIVE",
]

let factoryTransferStates = [
    "CREATED", "APPROVED", "PENDING", "ASSIGNED", "LOADING",
    "DISPATCHED", "IN_TRANSIT", "ARRIVED", "RECEIVED", "CANCELLED", "REASSIGNED",
]

let factoryVehicleStates = ["READY", "AVAILABLE", "UNAVAILABLE"]

let factoryDriverDuty = ["ON_SHIFT", "OFF_SHIFT"]

func canonicalizeOrderStatus(_ status: String) -> String {
    switch status.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() {
    case "DISPATCHED": return "LOADED"
    case "EN_ROUTE": return "IN_TRANSIT"
    case "ARRIVING": return "ARRIVED"
    case "SHOP_CLOSED_PENDING": return "ARRIVED_SHOP_CLOSED"
    default: return status.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
    }
}

func incrementOrderStatusCount(_ counts: [String: Int], status: String) -> [String: Int] {
    var next = Dictionary(uniqueKeysWithValues: orderStatusFunnel.map { ($0, 0) })
    for (key, value) in counts {
        let normalized = canonicalizeOrderStatus(key)
        if next[normalized] != nil {
            next[normalized] = value
        }
    }
    let key = canonicalizeOrderStatus(status)
    if next[key] != nil {
        next[key, default: 0] += 1
    }
    return next
}

func statusStackModel(
    dictionary: [String] = orderStatusFunnel,
    counts: [String: Int]?,
    available: Bool = true
) -> StatusStackModel {
    if !available {
        return StatusStackModel(
            mode: .unavailable,
            rows: dictionary.map { StatusStackRow(key: $0, count: nil, share: 0) },
            total: 0
        )
    }
    guard let counts else {
        return StatusStackModel(mode: .empty, rows: [], total: 0)
    }
    var rows = dictionary.map { key in
        StatusStackRow(key: key, count: counts[key] ?? 0, share: 0)
    }
    let total = rows.reduce(0) { $0 + ($1.count ?? 0) }
    if total > 0 {
        rows = rows.map { row in
            StatusStackRow(key: row.key, count: row.count, share: Double(row.count ?? 0) / Double(total))
        }
    }
    return StatusStackModel(mode: total == 0 ? .zero : .live, rows: rows, total: total)
}

struct SourceChip: View {
    let source: String

    var body: some View {
        Text(source.uppercased())
            .font(.caption2.weight(.semibold))
            .padding(.horizontal, 8)
            .padding(.vertical, 2)
            .overlay(Capsule().strokeBorder(.secondary, lineWidth: 1))
            .accessibilityIdentifier("gs-u-source-chip")
    }
}

struct CommandStatusJump: Hashable, Identifiable {
    let id: UUID
    let status: String
    var supplierId: String?

    init(status: String, supplierId: String? = nil) {
        self.id = UUID()
        self.status = canonicalizeOrderStatus(status)
        self.supplierId = supplierId
    }
}

func retailerOrderMatchesCommand(
    statusRaw: String,
    supplierId: String?,
    commandStatus: String,
    commandSupplierId: String?
) -> Bool {
    if canonicalizeOrderStatus(statusRaw) != canonicalizeOrderStatus(commandStatus) {
        return false
    }
    if let commandSupplierId, !commandSupplierId.isEmpty, supplierId != commandSupplierId {
        return false
    }
    return true
}

struct StatusStackView: View {
    let dictionary: [String]
    let counts: [String: Int]?
    var available: Bool = true
    var source: String? = nil
    var onSelect: ((String) -> Void)? = nil

    private var model: StatusStackModel {
        statusStackModel(dictionary: dictionary, counts: counts, available: available)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Spacer()
                SourceChip(source: source ?? (model.mode == .unavailable ? "unavailable" : model.mode == .empty ? "empty" : "live"))
            }
            if model.mode == .empty {
                Text("No status counts")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            } else if model.mode == .unavailable {
                Text("Status counts unavailable")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
            if model.mode == .live {
                GeometryReader { geo in
                    HStack(spacing: 0) {
                        ForEach(model.rows.filter { $0.share > 0 }, id: \.key) { row in
                            Rectangle()
                                .fill(Color.primary.opacity(0.35 + row.share * 0.65))
                                .frame(width: geo.size.width * row.share)
                        }
                    }
                }
                .frame(height: 10)
                .clipShape(Capsule())
            }
            if model.mode != .empty {
                FlexibleStatusChips(rows: model.rows, onSelect: onSelect)
            }
        }
        .accessibilityIdentifier("gs-u-status-stack")
    }
}

private struct FlexibleStatusChips: View {
    let rows: [StatusStackRow]
    var onSelect: ((String) -> Void)? = nil

    var body: some View {
        FlowLayout {
            ForEach(rows, id: \.key) { row in
                chip(row)
            }
        }
    }

    @ViewBuilder
    private func chip(_ row: StatusStackRow) -> some View {
        let label = HStack(spacing: 6) {
            Text(row.key.replacingOccurrences(of: "_", with: " "))
                .font(.caption2)
                .foregroundStyle(.secondary)
            Text(row.count.map(String.init) ?? "—")
                .font(.caption.monospacedDigit())
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .frame(minHeight: 44)
        .overlay(RoundedRectangle(cornerRadius: 8).strokeBorder(.secondary.opacity(0.4)))
        .accessibilityIdentifier("gs-u-chip-\(row.key)")

        if let onSelect {
            Button {
                onSelect(row.key)
            } label: {
                label
            }
            .buttonStyle(.plain)
        } else {
            label
        }
    }
}

File: pegasusX/packages/mobile-android-design/src/main/java/com/pegasus/design/ui/StatusStack.kt
package com.pegasus.design.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.unit.dp

enum class StatusStackMode { Empty, Zero, Live, Unavailable }

data class StatusStackRow(val key: String, val count: Int?, val share: Float)

data class StatusStackModel(
    val mode: StatusStackMode,
    val rows: List<StatusStackRow>,
    val total: Int,
)

val ORDER_STATUS_FUNNEL = listOf(
    "PENDING", "SCHEDULED", "AUTO_ACCEPTED", "BACKORDERED",
    "LOADED", "IN_TRANSIT", "DELAYED",
    "ARRIVED", "ARRIVED_SHOP_CLOSED",
    "AWAITING_PAYMENT", "PENDING_CASH_COLLECTION", "DELIVERED_ON_CREDIT",
    "FISCALIZING", "FISCAL_FAILED", "RECONCILIATION_REQUIRED",
    "COMPLETED", "CANCELLED",
)

val MANIFEST_STATES = listOf(
    "DRAFT", "LOADING", "SEALED", "DISPATCHED", "COMPLETED", "CANCELLED",
)

val TRUCK_DUTY_STATUSES = listOf(
    "AVAILABLE",
    "IN_TRANSIT",
    "RETURNING_TO_WAREHOUSE",
    "OFF_SHIFT",
    "UNASSIGNED",
    "VEHICLE_INACTIVE",
    "UNAVAILABLE",
    "INACTIVE",
)

val FACTORY_TRANSFER_STATES = listOf(
    "CREATED", "APPROVED", "PENDING", "ASSIGNED", "LOADING",
    "DISPATCHED", "IN_TRANSIT", "ARRIVED", "RECEIVED", "CANCELLED", "REASSIGNED",
)

val FACTORY_VEHICLE_STATES = listOf("READY", "AVAILABLE", "UNAVAILABLE")

val FACTORY_DRIVER_DUTY = listOf("ON_SHIFT", "OFF_SHIFT")

fun canonicalizeOrderStatus(status: String): String {
    return when (status.trim().uppercase()) {
        "DISPATCHED" -> "LOADED"
        "EN_ROUTE" -> "IN_TRANSIT"
        "ARRIVING" -> "ARRIVED"
        "SHOP_CLOSED_PENDING" -> "ARRIVED_SHOP_CLOSED"
        else -> status.trim().uppercase()
    }
}

fun incrementOrderStatusCount(counts: Map<String, Int>?, status: String): Map<String, Int> {
    val next = ORDER_STATUS_FUNNEL.associateWith { 0 }.toMutableMap()
    for ((key, value) in counts.orEmpty()) {
        val normalized = canonicalizeOrderStatus(key)
        if (normalized in next) {
            next[normalized] = value
        }
    }
    val key = canonicalizeOrderStatus(status)
    if (key in next) {
        next[key] = (next[key] ?: 0) + 1
    }
    return next
}

fun statusStackModel(
    dictionary: List<String> = ORDER_STATUS_FUNNEL,
    counts: Map<String, Int>?,
    available: Boolean = true,
): StatusStackModel {
    if (!available) {
        return StatusStackModel(
            mode = StatusStackMode.Unavailable,
            rows = dictionary.map { StatusStackRow(it, null, 0f) },
            total = 0,
        )
    }
    if (counts == null) {
        return StatusStackModel(StatusStackMode.Empty, emptyList(), 0)
    }
    val rows = dictionary.map { key -> StatusStackRow(key, counts[key] ?: 0, 0f) }.toMutableList()
    val total = rows.sumOf { it.count ?: 0 }
    if (total > 0) {
        for (i in rows.indices) {
            rows[i] = rows[i].copy(share = (rows[i].count ?: 0).toFloat() / total.toFloat())
        }
    }
    return StatusStackModel(
        mode = if (total == 0) StatusStackMode.Zero else StatusStackMode.Live,
        rows = rows,
        total = total,
    )
}

@Composable
fun SourceChip(source: String, modifier: Modifier = Modifier) {
    Text(
        text = source.uppercase(),
        style = MaterialTheme.typography.labelSmall,
        modifier = modifier
            .border(1.dp, MaterialTheme.colorScheme.outline, RoundedCornerShape(50))
            .padding(horizontal = 8.dp, vertical = 2.dp)
            .testTag("gs-u-source-chip"),
    )
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
fun StatusStack(
    counts: Map<String, Int>?,
    modifier: Modifier = Modifier,
    dictionary: List<String> = ORDER_STATUS_FUNNEL,
    available: Boolean = true,
    source: String? = null,
    onSelect: ((String) -> Unit)? = null,
) {
    val model = statusStackModel(dictionary, counts, available)
    val chipSource = source
        ?: when (model.mode) {
            StatusStackMode.Unavailable -> "unavailable"
            StatusStackMode.Empty -> "empty"
            else -> "live"
        }
    Column(modifier = modifier.testTag("gs-u-status-stack"), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            SourceChip(chipSource)
        }
        when (model.mode) {
            StatusStackMode.Empty -> Text("No status counts", style = MaterialTheme.typography.bodySmall)
            StatusStackMode.Unavailable -> Text("Status counts unavailable", style = MaterialTheme.typography.bodySmall)
            else -> Unit
        }
        if (model.mode == StatusStackMode.Live) {
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(10.dp)
                    .clip(RoundedCornerShape(50))
                    .background(MaterialTheme.colorScheme.surfaceVariant),
            ) {
                model.rows.filter { it.share > 0f }.forEach { row ->
                    Box(
                        Modifier
                            .fillMaxHeight()
                            .weight(row.share)
                            .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.35f + row.share * 0.65f)),
                    )
                }
            }
        }
        if (model.mode != StatusStackMode.Empty) {
            FlowRow(horizontalArrangement = Arrangement.spacedBy(8.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                model.rows.forEach { row ->
                    Row(
                        modifier = Modifier
                            .heightIn(min = 48.dp)
                            .border(1.dp, MaterialTheme.colorScheme.outline, RoundedCornerShape(8.dp))
                            .clickable(
                                enabled = onSelect != null,
                                role = Role.Button,
                                onClick = { onSelect?.invoke(row.key) },
                            )
                            .padding(horizontal = 8.dp, vertical = 4.dp)
                            .testTag("gs-u-chip-${row.key}"),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(6.dp),
                    ) {
                        Text(row.key.replace('_', ' '), style = MaterialTheme.typography.labelSmall)
                        Text(row.count?.toString() ?: "—", style = MaterialTheme.typography.labelMedium)
                    }
                }
            }
        }
    }
}
